### PR TITLE
Alpha version

### PR DESCRIPTION
**Pull request by young-n-rich**
Alpha version

### CODE REVIEW

File: src/bfast.h [hunk 1]
```diff
@@ -350,10 +350,10 @@ namespace bfast
 
         void write_file(string file) {
             auto data = pack();
-            FILE* f = nullptr;
-            if (fopen_s(&f, file.c_str(), "wb") != 0)
-                throw std::runtime_error("Failed to open file");
+            FILE* f = fopen(file.c_str(), "wb");
```
Comment (friesofdoom) [inline src/bfast.h]: Why was this changed for the less secure fopen?

File: src/bfast.h [hunk 2]
```diff
@@ -350,10 +350,10 @@ namespace bfast
...
             if (f == nullptr)
+            {
+            }
```
Comment (friesofdoom) [inline src/bfast.h]: This doesn't look right, the if statement does nothing now, and then the return below is always executed. Execution never reaches the fwrite and fclose below.

File: src/vim.h [hunk 3]
```diff
@@ -240,14 +240,14 @@ namespace Vim
                 }
                 else if (b.name == "nodes")
                 {
-                    mNodes = std::move(std::vector<SceneNode>((SceneNode*)b.data.begin(), (SceneNode*)b.data.end()));
+                    mNodes = std::vector<SceneNode>((SceneNode*)b.data.begin(), (SceneNode*)b.data.end());
```
Comment (friesofdoom) [inline src/vim.h:243]: Why was this changed, it looks like a reversal of a loading performance optimization.

File: src/vim.h [hunk 4]
```diff
@@ -240,14 +240,14 @@ namespace Vim
...
                 }
                 else if (b.name == "geometry")
                 {
                     try
                     {
                         mGeometryBFast = bfast::Bfast::unpack(b.data);
-                        mGeometry = std::move(g3d::G3d(mGeometryBFast));
+                        mGeometry = g3d::G3d(mGeometryBFast);
```
Comment (friesofdoom) [inline src/vim.h:250]: Same question, why was the move removed?

File: reference/Utils.cpp [hunk 5]
```diff
@@ -3,22 +3,23 @@
...
         struct stat stat_buf;
         int rc = stat(fileName.c_str(), &stat_buf);
         if (rc == 0)
         {
-            int fileSize = stat_buf.st_size;
+            off_t fileSize = stat_buf.st_size;
 
             FILE* file;
-            fopen_s(&file, fileName.c_str(), "rb");
+			file = fopen(fileName.c_str(), "rb");
```
Comment (friesofdoom) [inline reference/Utils.cpp:22]: why change from the more secure fopen_s to the less secure fopen?

File: VimToDatasmith/VimToDatasmith.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+//  VimToDatasmith
+//
+//  Created by Richard Young on 2021-05-17.
+//
+
+#pragma once
+
+#include "VimToDsWarningsDisabler.h"
+
+#define BeginVim2DatasmithNameSpace namespace Vim2Ds {
```
Comment (friesofdoom) [inline VimToDatasmith/VimToDatasmith.h]: This patterns seems to be more obfuscation than useful to me. Is there a reason to do it this way?

File: VimToDatasmith/VimToDatasmith.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+FVector GetTranslationVector(const cMat4& inMatrix)
+{
+	return FVector(inMatrix.m03 * Meter2Centimeter, -inMatrix.m13 * Meter2Centimeter, inMatrix.m23 * Meter2Centimeter);
+}
+
+
+// Extract scale from the matrix and return an Unreal one (in centimeters)
+FVector GetScaleVector(const cMat4& inMatrix)
+{
+	return FVector(sqrt(sqr(inMatrix.m00) + sqr(inMatrix.m10) + sqr(inMatrix.m20)),
```
Comment (friesofdoom) [inline VimToDatasmith/VimToDatasmith.cpp]: Is the matrix stored in column major? You could transpose the matrix and then use inMatrix.mRowX.Length(), but either way is fine really...

File: VimToDatasmith/VimToDatasmith.cpp [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+	RotAxis.Normalise();
+	
+	return FQuat(FVector(RotAxis.x, -RotAxis.y, RotAxis.z), RotAngle).Inverse();
+}
+
+
+// Extract translation from the matrix and return an Unreal one (in centimeters)
+FVector GetTranslationVector(const cMat4& inMatrix)
+{
+	return FVector(inMatrix.m03 * Meter2Centimeter, -inMatrix.m13 * Meter2Centimeter, inMatrix.m23 * Meter2Centimeter);
```
Comment (friesofdoom) [inline VimToDatasmith/VimToDatasmith.cpp]: This implies a handedness swap. Just letting you know.

File: VimToDatasmith/VimToDatasmith.cpp [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+bool IsIdentity(const cMat4& inTransform)
+{
+	return IsNearZero(inTransform.m16[0] - 1.0f) && IsNearZero(inTransform.m16[1]) && IsNearZero(inTransform.m16[2]) && IsNearZero(inTransform.m16[3]) &&
+	IsNearZero(inTransform.m16[4]) && IsNearZero(inTransform.m16[5] - 1.0f) && IsNearZero(inTransform.m16[6]) && IsNearZero(inTransform.m16[7]) &&
+	IsNearZero(inTransform.m16[8]) && IsNearZero(inTransform.m16[9]) && IsNearZero(inTransform.m16[10] - 1.0f) && IsNearZero(inTransform.m16[11]) &&
+	IsNearZero(inTransform.m16[12]) && IsNearZero(inTransform.m16[13]) && IsNearZero(inTransform.m16[14]) && IsNearZero(inTransform.m16[15] - 1.0f);
+}
+
+// Extract the rotation from the matrix and return as a Quat
+FQuat GetRotationQuat(const cMat4& inMatrix)
```
Comment (friesofdoom) [inline VimToDatasmith/VimToDatasmith.cpp]: You could possible use cQuat::FromMat4 in cQuat.h here

File: VimToDatasmith/VimToDatasmith.cpp [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+	return Utf8StringFormat("{(%s), (%s)}", ToUtf8(inAxisAlignedBoundingBox.mMin), ToUtf8(inAxisAlignedBoundingBox.mMax));
+}
+
+
+// To convert from Vim meters to UE centimeter units
+const float Meter2Centimeter = 100.0f;
+
+
+// Return true if this is an identity matrix
+bool IsIdentity(const cMat4& inTransform)
```
Comment (friesofdoom) [inline VimToDatasmith/VimToDatasmith.cpp]: You could add this to the matrix class

File: VimToDatasmith/CVimToDatasmith.cpp [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+
+	return meshElement;
+}
+
+
+// In old vim files, the geometry is exported in world space, even when instanced, so we need to remove that world transform from the geometry
+void CVimToDatasmith::FixOldVimFileTransforms()
+{
+	if (mVimScene.mVersionMajor == 0 && mVimScene.mVersionMinor == 0 && mVimScene.mVersionPatch <= 200) {
+		std::vector<bool>	IsTransformed(mGroupVertexOffets.Count(), false);
```
Comment (friesofdoom) [inline VimToDatasmith/CVimToDatasmith.cpp]: use a lower case i -> isTransformed

**Review by friesofdoom — CHANGES_REQUESTED** [review]
All looks pretty good but my main concern is the removal of the std::move operations.

File: src/vim.h (hunk 3 above)
Comment (young-n-rich) [reply src/vim.h:243]: /Volumes/RYDev/Dev/VIMaecGIT/vim-datasmith/src/vim.h:243:30: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
                    mNodes = std::move(std::vector<SceneNode>((SceneNode*)b.data.begin(), (SceneNode*)b.data.end()));
                             ^
/Volumes/RYDev/Dev/VIMaecGIT/vim-datasmith/src/vim.h:243:30: note: remove std::move call here
                    mNodes = std::move(std::vector<SceneNode>((SceneNode*)b.data.begin(), (SceneNode*)b.data.end()));
                             ^~~~~~~~~~                                                                            ~
/Volumes/RYDev/Dev/VIMaecGIT/vim-datasmith/src/vim.h:250:37: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
                        mGeometry = std::move(g3d::G3d(mGeometryBFast));
                                    ^
/Volumes/RYDev/Dev/VIMaecGIT/vim-datasmith/src/vim.h:250:37: note: remove std::move call here
                        mGeometry = std::move(g3d::G3d(mGeometryBFast));
                                    ^~~~~~~~~~                        ~

File: README.md [hunk 12]
```diff
@@ -1,2 +1,24 @@
 # vim-datasmith
 A VIM to DataSmith converter. 
+
+Build requirements
+You must build the Datasmith SDK
+	With Git
+		For that you need to get Unreal Engine git repository. Branch 4.27
```
Comment (cdiggins) [inline README.md:7]:
```suggestion
		For that you need to get [Unreal Engine Git repository](https://www.unrealengine.com/en-US/ue4-on-github). Branch 4.2
```

File: README.md [hunk 13]
```diff
@@ -1,2 +1,24 @@
...
+			"Engine/Source/Programs/Enterprise/Datasmith/DatasmithSDK/DatasmithSDK.Target.cs"
+			"Engine/Source/Programs/Enterprise/Datasmith/DatasmithSDK/DatasmithSDK.Build.cs"
+		Run "Setup.command"
+		Run GenerateProjectFiles.command
+		Open the UE4 project and build scheme "DatasmithSDK"
+	With Perforce
+		Just use the force (a lots) and everything will goes well.
+
+In the file 
+	In Vim-Datasmith.xcconfig - Edit "DatasmithSDKFolder" for the real path of your builded SDK
```
Comment (cdiggins) [inline README.md:18]: What is an "xconfig" file? Is there a Visual Studio project, or a make file, if I am using Windows and VS?

File: README.md [hunk 14]
```diff
@@ -1,2 +1,24 @@
 # vim-datasmith
 A VIM to DataSmith converter. 
+
+Build requirements
```
Comment (cdiggins) [inline README.md]:
```suggestion
## Build requirements
```

File: Settings/DatasmithSDK.Build.cs [hunk 15]
```diff
@@ -0,0 +1,24 @@
+// Copyright Epic Games, Inc. All Rights Reserved.
```
Comment (cdiggins) [inline Settings/DatasmithSDK.Build.cs:1]: What is the license of the Epic games code?

File: Settings/DatasmithSDK.Build.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+using System.IO;
+
+namespace UnrealBuildTool.Rules
+{
+	public class DatasmithSDK : ModuleRules
```
Comment (cdiggins) [inline Settings/DatasmithSDK.Build.cs:6]: Can you change all indentation to 4 spaces (not tabs)?

File: Settings/DatasmithSDK.Target.cs [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+		bShouldCompileAsDLL = true;
+
+		bBuildDeveloperTools = false;
+		bUseMallocProfiler = false;
+		bBuildWithEditorOnlyData = true;
+		bCompileAgainstEngine = false;
+		bCompileAgainstCoreUObject = true;
+		bCompileICU = false;
+		bUsesSlate = false;
+//		bDisableDebugInfo = true;
```
Comment (cdiggins) [inline Settings/DatasmithSDK.Target.cs:30]: Should this be #defined?

File: README.md (hunk 13 above)
Comment (young-n-rich) [reply README.md:18]: I will do Windows project for next monday.

File: README.md (hunk 12 above)
Comment (young-n-rich) [reply README.md:7]: Must be at least 4.27 (4.26 and before not valid)

File: Settings/DatasmithSDK.Target.cs (hunk 17 above)
Comment (young-n-rich) [reply Settings/DatasmithSDK.Target.cs:30]: This file is a temporary change to standard UE setup. Normally we will not have to change-it to build. But...

File: Settings/DatasmithSDK.Build.cs (hunk 16 above)
Comment (young-n-rich) [reply Settings/DatasmithSDK.Build.cs:6]: For all files or just this one? Have you a clang format file?

File: VimToDatasmith/CVimToDatasmith.h [hunk 18]
```diff
@@ -0,0 +1,204 @@
+//
+//  CVimToDatasmith.h
+//  CVimToDatasmith
+//
+//  Created by Richard Young on 2021-05-17.
+//
```
Comment (cdiggins) [inline VimToDatasmith/CVimToDatasmith.h]: For headers throughout let's just use

```suggestion
// Copyright (c) 2021 VIM
// Licensed under the MIT License 1.0
```

File: Settings/DatasmithSDK.Build.cs (hunk 15 above)
Comment (young-n-rich) [reply Settings/DatasmithSDK.Build.cs:1]: This file is a temporary change to standard UE setup. Normally we will not have to change-it to build. But...

File: VimToDatasmith/CVimToDatasmith.h [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+protected:
+	size_t mCount = 0; // Return the number of elements
+	C* mData = nullptr; // Array of the elements
+};
+
+
+// Vector class that refers to attribute content.
+/* Doesn't copy the attribute's content */
```
Comment (cdiggins) [inline VimToDatasmith/CVimToDatasmith.h:56]: I suggest using a line comment here.

File: VimToDatasmith/VimToDatasmith.h (hunk 6 above)
Comment (young-n-rich) [reply VimToDatasmith/VimToDatasmith.h]: You prefer using directly namespace Vim2Ds ?

File: VimToDatasmith/CVimToDatasmith.h [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+	TAllocatedVector<uint32_t>	mGroupIndexCounts;
+	TAllocatedVector<cVec3>	mNormals;
+
+	// Datasmith scene and assets output path
+	TSharedPtr< IDatasmithScene > mDatasmithScene;
+	FString mOutputPath;
+};
+
+
+EndVim2DatasmithNameSpace
```
Comment (cdiggins) [inline VimToDatasmith/CVimToDatasmith.h]: I'm a bit curious why the usage of a macro instead of using namespace explicitly.

File: VimToDatasmith/DebugTools.cpp [hunk 21]
```diff
@@ -0,0 +1,174 @@
+//
```
Comment (cdiggins) [inline VimToDatasmith/DebugTools.cpp]: Same comment before regarding file header comments throughout.

File: VimToDatasmith/VimToDatasmith.h (hunk 6 above)
Comment (friesofdoom) [reply VimToDatasmith/VimToDatasmith.h]: Yeah, it's not often I've seen it being #defined like this

File: src/vim.h (hunk 3 above)
Comment (friesofdoom) [reply src/vim.h:243]: Oh yes, it makes no sense for temporary objects.

Comment (cdiggins) [reply src/vim.h:243]: Are you satisfied with the answer @friesofdoom ?

**Review by cdiggins — APPROVED** [review]
Great work, @young-n-rich, thanks a lot! I've made a few minor comments.

File: src/vim.h (hunk 3 above)
Comment (friesofdoom) [reply src/vim.h:243]: Yes, I didn't notice the fact it was trying to move temporary objects, which doesn't make that much sense given RVO.

File: reference/Utils.cpp (hunk 5 above)
Comment (young-n-rich) [reply reference/Utils.cpp:22]: fopen_s is not available on macOS. Even if was part of C11, it is optional as other _s functions.
Anyway I will have to change it even on Windows for _wfopen_s that will permit to use unicode path.

File: Settings/DatasmithSDK.Build.cs (hunk 16 above)
Comment (cdiggins) [reply Settings/DatasmithSDK.Build.cs:6]: All files, no I don't have a clang format file. I only use VS.

File: src/bfast.h (hunk 1 above)
Comment (young-n-rich) [reply src/bfast.h]: I do this:
#ifndef __clang__
			FILE* f = nullptr;
            if (fopen_s(&f, file.c_str(), "wb") != 0)
                throw std::runtime_error("Failed to open file");
            if (f == nullptr)
                return;
#else
			// On macOS fopen_s is unavailable
			FILE* f = fopen(file.c_str(), "wb");
			if (f == nullptr)
				throw std::runtime_error("Failed to open file");
#endif

File: reference/Utils.cpp (hunk 5 above)
Comment (young-n-rich) [reply reference/Utils.cpp:22]:
```#ifndef __clang__
			FILE* f = nullptr;
            if (fopen_s(&f, file.c_str(), "wb") != 0)
                throw std::runtime_error("Failed to open file");
            if (f == nullptr)
                return;
#else
			// On macOS fopen_s is unavailable
			FILE* f = fopen(file.c_str(), "wb");
			if (f == nullptr)
				throw std::runtime_error("Failed to open file");
#endif
```

File: VimToDatasmith/VimToDatasmith.cpp (hunk 10 above)
Comment (young-n-rich) [reply VimToDatasmith/VimToDatasmith.cpp]: In don't use it now, I removed-it.

**Comment by cdiggins** [comment]
Great work, thanks!
